File: r/R/withings.R
# Withings API provider
#
# Wraps https://wbsapi.withings.net/ - uses TokenBridge for auth,
# handles pagination and date conversion internally.
#
# Usage:
#
#   tb_set_provider("withings")
#   tb_fetch("p001", "sleep-summary", start, end)   # canonical
#
#   wt_fetch("p001", "activity", start, end)         # Withings shorthand
#
#   tok <- tb_get_token("p001", provider = "withings")
#   wt_fetch("p001", "sleep-summary", start, end, token = tok)
#   wt_fetch("p001", "weight",        start, end, token = tok)
#
# Data type IDs: see names(WT_DATA_TYPES) or docs/providers/index.md
#
# Function prefix: wt_   (Withings)

# ── Data types reference ───────────────────────────────────────────────────────

#' All supported Withings data type IDs
#'
#' Named list describing fetch configuration for each type.
#' Use \code{names(WT_DATA_TYPES)} to list all available type IDs.
#'
#' Full descriptions and field details: see
#' \url{https://developer.withings.com/api-reference/}
#'
#' @export
#  Withings only returns a minimal default field set unless data_fields is
#  explicitly passed - full lists per endpoint (v2/sleep API reference), so
#  sleep-summary/sleep-detail return every available metric rather than
#  silently omitting AHI, HRV, respiration rate, etc.
.WT_SLEEP_GET_FIELDS <- paste(c(
  "hr", "rr", "snoring", "sdnn_1", "rmssd", "hrv_quality", "mvt_score",
  "chest_movement_rate", "withings_index", "breathing_sounds"
), collapse = ",")

.WT_SLEEP_SUMMARY_FIELDS <- paste(c(
  "total_timeinbed", "total_sleep_time", "asleepduration", "lightsleepduration",
  "remsleepduration", "deepsleepduration", "sleep_efficiency", "sleep_latency",
  "wakeup_latency", "wakeupduration", "wakeupcount", "waso", "nb_rem_episodes",
  "breathing_disturbances_intensity", "apnea_hypopnea_index", "withings_index",
  "durationtosleep", "durationtowakeup", "out_of_bed_count", "hr_average", "hr_min",
  "hr_max", "rr_average", "rr_min", "rr_max", "breathing_quality_assessment", "snoring",
  "snoringepisodecount", "sleep_score", "night_events", "mvt_score_avg",
  "mvt_active_duration", "rmssd_start_avg", "rmssd_end_avg",
  "chest_movement_rate_wellness_average", "chest_movement_rate_wellness_min",
  "chest_movement_rate_wellness_max", "breathing_sounds",
  "breathing_sounds_episode_count", "chest_movement_rate_average",
  "chest_movement_rate_min", "chest_movement_rate_max",
  "core_body_temperature_min", "core_body_temperature_max",
  "core_body_temperature_avg", "core_body_temperature_status"
), collapse = ",")

WT_DATA_TYPES <- list(
  # Sleep
  "sleep-summary" = list(endpoint = "/v2/sleep",   action = "getsummary", date_fmt = "ymd",  result_key = "series", data_fields = .WT_SLEEP_SUMMARY_FIELDS),
  "sleep-detail"  = list(endpoint = "/v2/sleep",   action = "get",        date_fmt = "unix", result_key = "series", data_fields = .WT_SLEEP_GET_FIELDS),
  # Activity
  "activity"      = list(endpoint = "/v2/measure", action = "getactivity",date_fmt = "ymd",  result_key = "activities"),
  "workouts"      = list(endpoint = "/v2/measure", action = "getworkouts",date_fmt = "ymd",  result_key = "series"),
  # Body measurements (via /measure getmeas)
  "weight"         = list(endpoint = "/measure", action = "getmeas", date_fmt = "unix", result_key = "measuregrps", meastype = 1L),
  "height"         = list(endpoint = "/measure", action = "getmeas", date_fmt = "unix", result_key = "measuregrps", meastype = 4L),
  "fat-ratio"      = list(endpoint = "/measure", action = "getmeas", date_fmt = "unix", result_key = "measuregrps", meastype = 6L),
  "fat-mass"       = list(endpoint = "/measure", action = "getmeas", date_fmt = "unix", result_key = "measuregrps", meastype = 8L),
  "blood-pressure" = list(endpoint = "/measure", action = "getmeas", date_fmt = "unix", result_key = "measuregrps", meastypes = "9,10"),
  "heart-rate"     = list(endpoint = "/measure", action = "getmeas", date_fmt = "unix", result_key = "measuregrps", meastype = 11L),
  "spo2"           = list(endpoint = "/measure", action = "getmeas", date_fmt = "unix", result_key = "measuregrps", meastype = 54L),
  "muscle-mass"    = list(endpoint = "/measure", action = "getmeas", date_fmt = "unix", result_key = "measuregrps", meastype = 76L),
  "bone-mass"      = list(endpoint = "/measure", action = "getmeas", date_fmt = "unix", result_key = "measuregrps", meastype = 88L),
  # ECG
  "ecg"           = list(endpoint = "/v2/heart", action = "list", date_fmt = "unix", result_key = "series")
)

# Human-readable labels for measure types (used in decode step)
.WT_MEASTYPE_LABELS <- c(
  "1"  = "weight_kg",
  "4"  = "height_m",
  "5"  = "fat_free_mass_kg",
  "6"  = "fat_ratio_pct",
  "8"  = "fat_mass_kg",
  "9"  = "diastolic_bp_mmhg",
  "10" = "systolic_bp_mmhg",
  "11" = "heart_rate_bpm",
  "12" = "temperature_c",
  "54" = "spo2_pct",
  "76" = "muscle_mass_kg",
  "88" = "bone_mass_kg"
)

# ── Public API ─────────────────────────────────────────────────────────────────

#' Fetch Withings data for a participant
#'
#' @param user_id    TokenBridge user ID
#' @param data_type  Data type ID (see \code{names(WT_DATA_TYPES)})
#' @param start_date "YYYY-MM-DD"
#' @param end_date   "YYYY-MM-DD"
#' @param token      Pre-fetched access token (skips TokenBridge round-trip)
#' @param sleepscan  If \code{TRUE}, resolve the Withings token via SleepScan using
#'   \code{user_id} as the lookup key (email, Withings user ID as integer, or
#'   SleepScan participant ID). Takes priority over \code{token=}.
#'   Requires \code{SLEEPSCAN_API_KEY} in the environment or \code{sleepscan_key=}.
#' @param sleepscan_key  SleepScan API key. Falls back to \code{SLEEPSCAN_API_KEY} env var.
#' @param env_file   Path to .env file (default ".env")
#' @return data.frame, one row per record
#' @export
wt_fetch <- function(user_id, data_type, start_date, end_date,
                     token         = NULL,
                     sleepscan     = NULL,
                     sleepscan_key = NULL,
                     env_file      = ".env") {
  .wt_validate_dates(start_date, end_date)

  if (!data_type %in% names(WT_DATA_TYPES)) {
    stop(
      "Unknown Withings data type: ", shQuote(data_type),
      ". Available: ", paste(names(WT_DATA_TYPES), collapse = ", "),
      call. = FALSE
    )
  }

  if (isTRUE(sleepscan)) {
    token <- .wt_sleepscan_token(user_id, sleepscan_key)
  } else if (is.null(token)) {
    token <- tb_get_token(user_id, provider = "withings", env_file = env_file)
  }

  .wt_fetch_all(token, data_type, start_date, end_date)
}

# ── Internal ───────────────────────────────────────────────────────────────────

.wt_sleepscan_token <- function(sleepscan, sleepscan_key = NULL) {
  key <- sleepscan_key %||% Sys.getenv("SLEEPSCAN_API_KEY")
  if (!nchar(key))
    stop("sleepscan= requires SLEEPSCAN_API_KEY in environment or sleepscan_key= argument.",
         call. = FALSE)
  client <- ss_client(key)
  if (is.numeric(sleepscan)) {
    ss_get_token(client, withings_user_id = as.integer(sleepscan))
  } else if (grepl("@", sleepscan, fixed = TRUE)) {
    ss_get_token(client, email = sleepscan)
  } else {
    ss_get_token(client, participant_id = sleepscan)
  }
}

.wt_validate_dates <- function(start_date, end_date) {
  fmt <- "%Y-%m-%d"
  s <- tryCatch(as.Date(start_date, format = fmt), warning = function(w) NA)
  e <- tryCatch(as.Date(end_date,   format = fmt), warning = function(w) NA)

  if (is.na(s) || is.na(e) ||
      format(s, fmt) != start_date || format(e, fmt) != end_date) {
    stop("Dates must be YYYY-MM-DD format, got: ",
         shQuote(start_date), ", ", shQuote(end_date), call. = FALSE)
  }
  if (s > e) {
    stop("start_date (", start_date, ") must not be after end_date (", end_date, ")",
         call. = FALSE)
  }
}

.wt_to_unix <- function(date_str, end_of_day = FALSE) {
  d <- as.POSIXct(date_str, tz = "UTC")
  if (end_of_day) d <- d + 86399L  # 23:59:59
  as.integer(d)
}

.wt_decode_measuregrps <- function(grps) {
  rows <- lapply(grps, function(grp) {
    row <- list(
      grpid    = grp$grpid,
      date     = grp$date,
      category = grp$category
    )
    for (m in grp$measures) {
      actual_val <- m$value * (10 ^ m$unit)
      col <- .WT_MEASTYPE_LABELS[as.character(m$type)]
      col <- if (is.na(col)) paste0("type_", m$type) else col
      row[[col]] <- round(actual_val, 6)
    }
    row
  })
  if (length(rows) == 0L) return(data.frame())
  do.call(dplyr_bind_rows_fallback, list(rows))
}

# Fallback row-binder that doesn't require dplyr
dplyr_bind_rows_fallback <- function(rows) {
  all_keys <- unique(unlist(lapply(rows, names)))
  mat <- lapply(rows, function(r) {
    lapply(setNames(all_keys, all_keys), function(k) {
      if (!is.null(r[[k]])) r[[k]] else NA
    })
  })
  as.data.frame(do.call(rbind, lapply(mat, as.data.frame, stringsAsFactors = FALSE)),
                stringsAsFactors = FALSE)
}

.wt_fetch_all <- function(token, data_type, start_date, end_date) {
  spec       <- WT_DATA_TYPES[[data_type]]
  base_url   <- "https://wbsapi.withings.net"
  url        <- paste0(base_url, spec$endpoint)
  result_key <- spec$result_key
  headers    <- httr::add_headers(Authorization = paste("Bearer", token))

  # Date params
  if (spec$date_fmt == "ymd") {
    date_params <- list(startdateymd = start_date, enddateymd = end_date)
  } else {
    date_params <- list(
      startdate = .wt_to_unix(start_date),
      enddate   = .wt_to_unix(end_date, end_of_day = TRUE)
    )
  }

  # Extra params (meastype / meastypes for getmeas, data_fields for sleep)
  extra <- list()
  if (!is.null(spec$meastype))    extra$meastype    <- spec$meastype
  if (!is.null(spec$meastypes))   extra$meastypes   <- spec$meastypes
  if (!is.null(spec$data_fields)) extra$data_fields <- spec$data_fields

  all_records <- list()
  offset      <- NULL

  repeat {
    params <- c(list(action = spec$action), date_params, extra)
    if (!is.null(offset)) params$offset <- offset

    resp <- httr::GET(url, headers, query = params)

    if (httr::status_code(resp) != 200L) {
      stop("Withings API HTTP error: ", httr::status_code(resp), call. = FALSE)
    }

    body <- httr::content(resp, as = "parsed", type = "application/json")

    wt_status <- body$status %||% -1L
    if (wt_status != 0L) {
      stop("Withings API error (status ", wt_status, "): ",
           body$error %||% "unknown", call. = FALSE)
    }

    data <- body$body
    page_records <- data[[result_key]] %||% list()

    if (result_key == "measuregrps") {
      decoded <- .wt_decode_measuregrps(page_records)
      all_records <- c(all_records, list(decoded))
    } else {
      all_records <- c(all_records, page_records)
    }

    if (isTRUE(data$more)) {
      offset <- data$offset
    } else {
      break
    }
  }

  if (length(all_records) == 0L) return(data.frame())

  if (result_key == "measuregrps") {
    do.call(rbind, all_records)
  } else {
    as.data.frame(
      do.call(rbind, lapply(all_records, function(r) {
        as.data.frame(lapply(r, function(v) if (length(v) == 1) v else list(v)),
                      stringsAsFactors = FALSE)
      })),
      stringsAsFactors = FALSE
    )
  }
}

# Null-coalescing operator (if not already defined)
`%||%` <- function(a, b) if (!is.null(a)) a else b

File: python/tokenbridge/providers/withings.py
"""Withings Health API provider.

Wraps the Withings API (https://wbsapi.withings.net/) - uses TokenBridge for
auth, handles pagination and date conversion internally.

All responses are returned as flat lists of dicts.  Nested fields are
flattened with dot notation where applicable.

Supported data types::

    from tokenbridge.providers.withings import DATA_TYPES
    print(list(DATA_TYPES))

See https://developer.withings.com/api-reference/ for field-level details.
"""

from datetime import date, datetime, timezone
from typing import Optional, Sequence

import requests

from tokenbridge.providers._base import HealthProvider

_BASE = "https://wbsapi.withings.net"

# Each entry describes how to fetch the type:
#   endpoint  - API path (relative to _BASE)
#   action    - value of the ?action= param
#   date_fmt  - "ymd" (YYYY-MM-DD strings) or "unix" (epoch seconds)
#   result_key - key inside response body containing the data list
#   meastype  - (getmeas only) single measure type int
#   meastypes - (getmeas only) comma-separated type ints string
DATA_TYPES: dict[str, dict] = {
    # ── Sleep ────────────────────────────────────────────────────────────────
    "sleep-summary": {
        "endpoint": "/v2/sleep", "action": "getsummary",
        "date_fmt": "ymd", "result_key": "series",
    },
    "sleep-detail": {
        "endpoint": "/v2/sleep", "action": "get",
        "date_fmt": "unix", "result_key": "series",
    },
    # ── Activity ─────────────────────────────────────────────────────────────
    "activity": {
        "endpoint": "/v2/measure", "action": "getactivity",
        "date_fmt": "ymd", "result_key": "activities",
    },
    "workouts": {
        "endpoint": "/v2/measure", "action": "getworkouts",
        "date_fmt": "ymd", "result_key": "series",
    },
    # ── Body measurements ─────────────────────────────────────────────────────
    "weight":         {"endpoint": "/measure", "action": "getmeas", "date_fmt": "unix", "result_key": "measuregrps", "meastype": 1},
    "height":         {"endpoint": "/measure", "action": "getmeas", "date_fmt": "unix", "result_key": "measuregrps", "meastype": 4},
    "fat-ratio":      {"endpoint": "/measure", "action": "getmeas", "date_fmt": "unix", "result_key": "measuregrps", "meastype": 6},
    "fat-mass":       {"endpoint": "/measure", "action": "getmeas", "date_fmt": "unix", "result_key": "measuregrps", "meastype": 8},
    "blood-pressure": {"endpoint": "/measure", "action": "getmeas", "date_fmt": "unix", "result_key": "measuregrps", "meastypes": "9,10"},
    "heart-rate":     {"endpoint": "/measure", "action": "getmeas", "date_fmt": "unix", "result_key": "measuregrps", "meastype": 11},
    "spo2":           {"endpoint": "/measure", "action": "getmeas", "date_fmt": "unix", "result_key": "measuregrps", "meastype": 54},
    "muscle-mass":    {"endpoint": "/measure", "action": "getmeas", "date_fmt": "unix", "result_key": "measuregrps", "meastype": 76},
    "bone-mass":      {"endpoint": "/measure", "action": "getmeas", "date_fmt": "unix", "result_key": "measuregrps", "meastype": 88},
    # ── Heart / ECG ──────────────────────────────────────────────────────────
    "ecg": {
        "endpoint": "/v2/heart", "action": "list",
        "date_fmt": "unix", "result_key": "series",
    },
}

# Measure types that decode a single numeric value (value * 10^unit)
_SCALAR_MEASTYPES = {
    1: "weight_kg",
    4: "height_m",
    5: "fat_free_mass_kg",
    6: "fat_ratio_pct",
    8: "fat_mass_kg",
    9: "diastolic_bp_mmhg",
    10: "systolic_bp_mmhg",
    11: "heart_rate_bpm",
    12: "temperature_c",
    54: "spo2_pct",
    71: "body_temperature_c",
    73: "skin_temperature_c",
    76: "muscle_mass_kg",
    77: "hydration_kg",
    88: "bone_mass_kg",
    91: "pulse_wave_velocity_m_s",
    123: "vo2max_ml_min_kg",
}


def _validate_dates(start_date: str, end_date: str) -> None:
    fmt = "%Y-%m-%d"
    try:
        s = datetime.strptime(start_date, fmt).date()
        e = datetime.strptime(end_date, fmt).date()
    except ValueError:
        raise ValueError(
            f"Dates must be YYYY-MM-DD format, got: {start_date!r}, {end_date!r}"
        )
    if s > e:
        raise ValueError(
            f"start_date ({start_date}) must not be after end_date ({end_date})"
        )


def _to_unix(date_str: str, end_of_day: bool = False) -> int:
    """Convert YYYY-MM-DD to a unix timestamp (UTC midnight, or 23:59:59)."""
    d = date.fromisoformat(date_str)
    if end_of_day:
        dt = datetime(d.year, d.month, d.day, 23, 59, 59, tzinfo=timezone.utc)
    else:
        dt = datetime(d.year, d.month, d.day, tzinfo=timezone.utc)
    return int(dt.timestamp())


def _decode_measuregrps(grps: list[dict]) -> list[dict]:
    """Flatten Withings measuregrps into one dict per measurement group."""
    rows = []
    for grp in grps:
        row: dict = {
            "grpid":    grp.get("grpid"),
            "date":     grp.get("date"),
            "category": grp.get("category"),
        }
        for m in grp.get("measures", []):
            mtype = m.get("type")
            raw_val = m.get("value", 0) * (10 ** m.get("unit", 0))
            col = _SCALAR_MEASTYPES.get(mtype, f"type_{mtype}")
            row[col] = round(raw_val, 6)
        rows.append(row)
    return rows


class Withings(HealthProvider):
    """Withings health data provider.

    Usually accessed via ``tb.withings`` rather than instantiated directly.

    All public methods accept an optional ``token`` argument.  Pass a
    pre-fetched token to avoid one TokenBridge round-trip per call:

    .. code-block:: python

        token = tb.get_token("p001", provider="withings")
        sleep = wt.fetch("p001", "sleep-summary", start, end, token=token)
        weight = wt.fetch("p001", "weight",       start, end, token=token)
    """

    PROVIDER_ID = "withings"

    def fetch(
        self,
        user_id: str,
        data_type: str,
        start_date: str,
        end_date: str,
        *,
        token: Optional[str] = None,
        raw: bool = False,
    ) -> list[dict]:
        """Fetch any Withings data type by its ID.

        Args:
            user_id: TokenBridge participant ID.
            data_type: Type ID from :data:`DATA_TYPES`, e.g. ``"sleep-summary"``,
                ``"activity"``, ``"weight"``.
            start_date: Start of date range, ``"YYYY-MM-DD"``.
            end_date: End of date range, ``"YYYY-MM-DD"``.
            token: Pre-fetched access token.  Obtain with
                ``tb.get_token(user_id, provider="withings")``.

        Returns:
            List of flat dicts, one per record.  Returns an empty list if no
            data exists for the period.

        Raises:
            ValueError: If ``data_type`` is not in :data:`DATA_TYPES` or dates
                are invalid.
            RuntimeError: If the Withings API returns a non-zero status.
        """
        _validate_dates(start_date, end_date)
        if data_type not in DATA_TYPES:
            raise ValueError(
                f"Unknown Withings data type {data_type!r}. "
                f"Available: {sorted(DATA_TYPES)}"
            )
        if token is None:
            token = self._get_token(user_id)
        return _fetch_all(token, data_type, start_date, end_date, raw=raw)


# ── Internal fetch helpers ────────────────────────────────────────────────────

def _fetch_all(token: str, data_type: str, start_date: str, end_date: str, raw: bool = False) -> list[dict]:
    """Fetch all pages for a data type and return combined records."""
    spec = DATA_TYPES[data_type]
    endpoint = spec["endpoint"]
    action = spec["action"]
    date_fmt = spec["date_fmt"]
    result_key = spec["result_key"]

    # Build date params
    if date_fmt == "ymd":
        date_params = {"startdateymd": start_date, "enddateymd": end_date}
    else:
        date_params = {
            "startdate": _to_unix(start_date),
            "enddate":   _to_unix(end_date, end_of_day=True),
        }

    # Extra params for getmeas
    extra: dict = {}
    if "meastype" in spec:
        extra["meastype"] = spec["meastype"]
    if "meastypes" in spec:
        extra["meastypes"] = spec["meastypes"]

    headers = {"Authorization": f"Bearer {token}"}
    url = f"{_BASE}{endpoint}"
    records: list[dict] = []
    offset: Optional[int] = None

    while True:
        params = {"action": action, **date_params, **extra}
        if offset is not None:
            params["offset"] = offset

        resp = requests.get(url, params=params, headers=headers, timeout=30)
        resp.raise_for_status()
        body = resp.json()

        status = body.get("status", -1)
        if status != 0:
            raise RuntimeError(
                f"Withings API error (status {status}) for {data_type!r}: "
                f"{body.get('error', body)}"
            )

        if raw:
            records.append(body)
            if not body.get("body", {}).get("more"):
                break
            offset = body["body"].get("offset")
            continue

        data = body.get("body", {})
        page_records = data.get(result_key, []) or []

        if result_key == "measuregrps":
            page_records = _decode_measuregrps(page_records)

        records.extend(page_records)

        if data.get("more"):
            offset = data.get("offset")
        else:
            break

    return records
